File: reference/python/zakai_settle.py
"""
Zakai Settlement — reference adjudication, Python, zero dependencies.

WHY THIS ONE MATTERS MORE THAN THE AUTHORIZATION REFERENCE

The decision layer compares values. This layer hashes records, and hashing is
where signed-record formats actually break between languages. Two
implementations that serialise the same object differently — a different key
order, a differently handled absent field, a float where the other has an int —
compute different hashes, reject each other's perfectly valid chains, and each
concludes the other's cryptography is at fault.

So the canonicalisation is written out explicitly below rather than left to
whatever `json.dumps` happens to do, and the published vectors carry hash
fixtures you should check before looking at a single verdict. An implementation
that produces the right verdicts from the wrong hashes has agreed with nobody:
it works on chains it built itself and fails on every chain built by anyone
else.

    python3 zakai_settle.py --file settlement-vectors.json
    python3 zakai_settle.py --url https://zakai-3uxj.vercel.app

Exit code is 0 when every vector and every hash fixture passes, 1 otherwise.
"""

from __future__ import annotations

import hashlib
import json
import sys
import urllib.request
from typing import Any

DEFAULT_ORIGIN = "https://zakai-3uxj.vercel.app"
VECTORS_PATH = "/api/settlement/test-vectors"

_UNSET = object()


def canonical(value: Any) -> str:
    """Serialise so two implementations produce identical bytes.

    Three rules, and each is a place implementations silently diverge:

      * object keys sorted by code unit, ascending
      * fields whose value is absent are omitted entirely, so an absent field
        and an explicitly-null-because-unset one hash identically
      * no insignificant whitespace; arrays keep their order

    Python's ``None`` is used for JSON ``null``, which is a real value and is
    kept. The omitted case is a key that is simply not present — this function
    never sees it — plus the sentinel below, which exists so a caller porting
    from a language with an explicit "undefined" has somewhere to put it.
    """
    if value is _UNSET:
        raise ValueError("unset values must be omitted, not serialised")
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, (int, float)):
        # Integers only, in practice: money is always minor units and never a
        # float. A float that is integral is emitted without a decimal point so
        # it matches a language that carried it as an integer.
        if isinstance(value, float) and value.is_integer():
            return str(int(value))
        return json.dumps(value)
    if isinstance(value, str):
        return json.dumps(value, ensure_ascii=False)
    if isinstance(value, (list, tuple)):
        return "[" + ",".join(canonical(v) for v in value) + "]"
    if isinstance(value, dict):
        items = sorted((k, v) for k, v in value.items() if v is not _UNSET)
        return "{" + ",".join(f"{json.dumps(k, ensure_ascii=False)}:{canonical(v)}" for k, v in items) + "}"
    raise TypeError(f"cannot canonicalise {type(value)!r}")


def hash_record(record: Any) -> str:
    return hashlib.sha256(canonical(record).encode("utf-8")).hexdigest()


def _verdict(verdict: str, burden: str, settled: int = 0) -> dict[str, Any]:
    return {"verdict": verdict, "burden": burden, "settledMinor": settled}


def adjudicate(chain: dict[str, Any], now: int) -> dict[str, Any]:
    """Who is right, from the records alone.

    Pure and total: every input yields a verdict and none raises. A function a
    bank wraps in try/except is one whose except block will eventually decide a
    dispute.
    """
    mandate = chain["mandate"]
    decision = chain.get("decision")
    outcome = chain.get("outcome")

    # An outcome with no decision before it is the case this layer exists for:
    # something was done and nobody can point at the permission for it.
    if outcome and not decision:
        return _verdict("unauthorized", "institution")
    if not decision:
        # A real verdict, not a failure. A procedure that always produces a
        # winner will sometimes invent one.
        return _verdict("indeterminate", "none")

    if decision.get("mandateJti") != mandate.get("jti"):
        return _verdict("broken_chain", "institution")
    if decision.get("prevHash") != mandate.get("hash"):
        return _verdict("broken_chain", "institution")

    if decision.get("decision") == "deny":
        # A refusal is a recorded answer, not a fault. Treating it as fault is
        # how a network punishes the participants who behave correctly.
        if outcome and outcome.get("result") != "refused":
            return _verdict("unauthorized", "institution")
        return _verdict("refused_with_reason", "none")

    at = decision.get("at")
    if at < mandate["nbf"] or at >= mandate["exp"]:
        return _verdict("outside_mandate_window", "institution")
    if decision.get("action") not in (mandate.get("scopes") or ()):
        return _verdict("exceeded_scope", "institution")

    if not outcome:
        # The verdict that makes silence expensive, which is the point: the
        # burden sits with the party holding the record of what it did.
        return _verdict("authorized_not_performed", "institution")

    if outcome.get("prevHash") != hash_record(decision):
        return _verdict("broken_chain", "institution")
    if outcome.get("action") != decision.get("action"):
        return _verdict("exceeded_scope", "institution")
    if outcome.get("at") < at:
        return _verdict("broken_chain", "institution")
    if outcome.get("at") >= mandate["exp"]:
        return _verdict("outside_mandate_window", "institution")

    if outcome.get("result") == "refused":
        return _verdict("refused_with_reason", "none")

    amount = outcome.get("amountMinor") or 0
    # Negative or fractional money is refused rather than coerced. A settlement
    # layer that quietly rounds is one whose totals stop reconciling, and the
    # first party to notice will be the one owed the rounding.
    if isinstance(amount, bool) or not isinstance(amount, int) or amount < 0:
        if not (isinstance(amount, float) and float(amount).is_integer() and amount >= 0):
            return _verdict("indeterminate", "institution")
        amount = int(amount)

    return _verdict("performed_as_authorized", "none", int(amount))


# ---------------------------------------------------------------------------
# Conformance runner
# ---------------------------------------------------------------------------


def load(source: str) -> dict[str, Any]:
    if not source.startswith(("http://", "https://")):
        with open(source, encoding="utf-8") as handle:
            return json.load(handle)
    with urllib.request.urlopen(source.rstrip("/") + VECTORS_PATH, timeout=30) as resp:
        return json.load(resp)


def main(argv: list[str]) -> int:
    source = DEFAULT_ORIGIN
    for i, arg in enumerate(argv[:-1]):
        if arg in ("--url", "--file"):
            source = argv[i + 1]

    try:
        doc = load(source)
    except Exception as err:  # noqa: BLE001
        print(f"could not load vectors from {source}: {err}", file=sys.stderr)
        return 2

    # Hashes first. A right verdict from a wrong hash is agreement about
    # nothing, so there is no point reading the verdicts until this passes.
    hash_failures = []
    for fixture in doc["canonicalisation"]["fixtures"]:
        got = hash_record(fixture["record"])
        if got != fixture["sha256"]:
            hash_failures.append(f"  canonical hash mismatch\n    expected {fixture['sha256']}\n    got      {got}")

    if hash_failures:
        print("NOT CONFORMANT - canonicalisation differs:\n")
        print("\n".join(hash_failures))
        return 1

    now = int(doc["evaluated_at_unix"])
    failures = []
    for v in doc["vectors"]:
        expect = v["expect"]
        expected = f"{expect['verdict']}/{expect['burden']}/{expect['settledMinor']}"
        try:
            got_obj = adjudicate(v["chain"], now)
            got = f"{got_obj['verdict']}/{got_obj['burden']}/{got_obj['settledMinor']}"
        except Exception as err:  # noqa: BLE001 - a throw is a failure, not a crash
            got = f"threw:{err}"
        if got != expected:
            failures.append(f"  {v['id']}: expected {expected}, got {got}\n    pins: {v['pins']}")

    total = len(doc["vectors"])
    if failures:
        # No partial credit: one wrong verdict in a settlement network is one
        # dispute resolved differently depending on who was asked.
        print(f"NOT CONFORMANT - {len(failures)} of {total} vectors failed:\n")
        print("\n".join(failures))
        return 1

    print(f"CONFORMANT - {total}/{total} vectors and {len(doc['canonicalisation']['fixtures'])} hash fixtures passed.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: reference/go/zakai_settle.go
// Zakai Settlement — reference adjudication, Go, standard library only.
//
// Companion to zakai_decide.go and the Python settlement reference. Banks and
// PSPs evaluate Go first; a settlement layer that only speaks TypeScript and
// Python is one a risk committee can dismiss as "vendor language."
//
// Canonicalisation is spelled out explicitly. Two implementations that
// serialise the same object differently compute different hashes and reject
// each other's valid chains. Verdict agreement from disagreeing hashes is
// agreement about nothing.
//
//	go run zakai_settle.go --url https://zakai-3uxj.vercel.app
//	go run zakai_settle.go --file settlement-vectors.json
//
// Exit 0 when every vector and every hash fixture passes, 1 otherwise.
package main

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
)

const (
	defaultOrigin = "https://zakai-3uxj.vercel.app"
	vectorsPath   = "/api/settlement/test-vectors"
)

// canonical serialises so two implementations produce identical bytes.
// Rules: object keys sorted by code unit; absent fields omitted; no
// insignificant whitespace; arrays keep order; integral floats emit as ints.
func canonical(value any) (string, error) {
	switch v := value.(type) {
	case nil:
		return "null", nil
	case bool:
		if v {
			return "true", nil
		}
		return "false", nil
	case json.Number:
		// Prefer integer form when the number is integral.
		if i, err := v.Int64(); err == nil {
			return strconv.FormatInt(i, 10), nil
		}
		f, err := v.Float64()
		if err != nil {
			return "", err
		}
		if f == float64(int64(f)) {
			return strconv.FormatInt(int64(f), 10), nil
		}
		b, err := json.Marshal(f)
		return string(b), err
	case float64:
		if v == float64(int64(v)) {
			return strconv.FormatInt(int64(v), 10), nil
		}
		b, err := json.Marshal(v)
		return string(b), err
	case int:
		return strconv.Itoa(v), nil
	case int64:
		return strconv.FormatInt(v, 10), nil
	case string:
		b, err := json.Marshal(v)
		return string(b), err
	case []any:
		parts := make([]string, len(v))
		for i, item := range v {
			s, err := canonical(item)
			if err != nil {
				return "", err
			}
			parts[i] = s
		}
		return "[" + strings.Join(parts, ",") + "]", nil
	case map[string]any:
		keys := make([]string, 0, len(v))
		for k := range v {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		parts := make([]string, 0, len(keys))
		for _, k := range keys {
			ks, err := json.Marshal(k)
			if err != nil {
				return "", err
			}
			vs, err := canonical(v[k])
			if err != nil {
				return "", err
			}
			parts = append(parts, string(ks)+":"+vs)
		}
		return "{" + strings.Join(parts, ",") + "}", nil
	default:
		return "", fmt.Errorf("cannot canonicalise %T", value)
	}
}

func hashRecord(record any) (string, error) {
	s, err := canonical(record)
	if err != nil {
		return "", err
	}
	sum := sha256.Sum256([]byte(s))
	return fmt.Sprintf("%x", sum), nil
}

type verdict struct {
	Verdict      string `json:"verdict"`
	Burden       string `json:"burden"`
	SettledMinor int64  `json:"settledMinor"`
}

func asMap(v any) map[string]any {
	m, _ := v.(map[string]any)
	return m
}

func asString(v any) string {
	s, _ := v.(string)
	return s
}

func asInt64(v any) (int64, bool) {
	switch n := v.(type) {
	case json.Number:
		i, err := n.Int64()
		return i, err == nil
	case float64:
		if n == float64(int64(n)) {
			return int64(n), true
		}
		return 0, false
	case int64:
		return n, true
	case int:
		return int64(n), true
	default:
		return 0, false
	}
}

func stringSlice(v any) []string {
	arr, ok := v.([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(arr))
	for _, item := range arr {
		if s, ok := item.(string); ok {
			out = append(out, s)
		}
	}
	return out
}

func containsString(hay []string, needle string) bool {
	for _, s := range hay {
		if s == needle {
			return true
		}
	}
	return false
}

func adjudicate(chain map[string]any, now int64) verdict {
	_ = now
	mandate := asMap(chain["mandate"])
	decision := asMap(chain["decision"])
	outcome := asMap(chain["outcome"])

	if outcome != nil && decision == nil {
		return verdict{Verdict: "unauthorized", Burden: "institution"}
	}
	if decision == nil {
		return verdict{Verdict: "indeterminate", Burden: "none"}
	}

	if asString(decision["mandateJti"]) != asString(mandate["jti"]) {
		return verdict{Verdict: "broken_chain", Burden: "institution"}
	}
	if asString(decision["prevHash"]) != asString(mandate["hash"]) {
		return verdict{Verdict: "broken_chain", Burden: "institution"}
	}

	if asString(decision["decision"]) == "deny" {
		if outcome != nil && asString(outcome["result"]) != "refused" {
			return verdict{Verdict: "unauthorized", Burden: "institution"}
		}
		return verdict{Verdict: "refused_with_reason", Burden: "none"}
	}

	at, atOk := asInt64(decision["at"])
	nbf, _ := asInt64(mandate["nbf"])
	exp, _ := asInt64(mandate["exp"])
	if !atOk || at < nbf || at >= exp {
		return verdict{Verdict: "outside_mandate_window", Burden: "institution"}
	}
	if !containsString(stringSlice(mandate["scopes"]), asString(decision["action"])) {
		return verdict{Verdict: "exceeded_scope", Burden: "institution"}
	}

	if outcome == nil {
		return verdict{Verdict: "authorized_not_performed", Burden: "institution"}
	}

	decHash, err := hashRecord(decision)
	if err != nil || asString(outcome["prevHash"]) != decHash {
		return verdict{Verdict: "broken_chain", Burden: "institution"}
	}
	if asString(outcome["action"]) != asString(decision["action"]) {
		return verdict{Verdict: "exceeded_scope", Burden: "institution"}
	}
	outAt, outOk := asInt64(outcome["at"])
	if !outOk || outAt < at {
		return verdict{Verdict: "broken_chain", Burden: "institution"}
	}
	if outAt >= exp {
		return verdict{Verdict: "outside_mandate_window", Burden: "institution"}
	}

	if asString(outcome["result"]) == "refused" {
		return verdict{Verdict: "refused_with_reason", Burden: "none"}
	}

	amountRaw := outcome["amountMinor"]
	if amountRaw == nil {
		amountRaw = float64(0)
	}
	amount, amountOk := asInt64(amountRaw)
	if !amountOk || amount < 0 {
		return verdict{Verdict: "indeterminate", Burden: "institution"}
	}

	return verdict{Verdict: "performed_as_authorized", Burden: "none", SettledMinor: amount}
}

type settleDoc struct {
	EvaluatedAtUnix int64 `json:"evaluated_at_unix"`
	Canonicalisation struct {
		Fixtures []struct {
			Record any    `json:"record"`
			SHA256 string `json:"sha256"`
		} `json:"fixtures"`
	} `json:"canonicalisation"`
	Vectors []struct {
		ID     string         `json:"id"`
		Pins   string         `json:"pins"`
		Chain  map[string]any `json:"chain"`
		Expect struct {
			Verdict      string `json:"verdict"`
			Burden       string `json:"burden"`
			SettledMinor int64  `json:"settledMinor"`
		} `json:"expect"`
	} `json:"vectors"`
}

func load(source string) (*settleDoc, error) {
	var r io.Reader
	if strings.HasPrefix(source, "http://") || strings.HasPrefix(source, "https://") {
		client := &http.Client{Timeout: 30 * time.Second}
		resp, err := client.Get(strings.TrimSuffix(source, "/") + vectorsPath)
		if err != nil {
			return nil, err
		}
		defer resp.Body.Close()
		r = resp.Body
	} else {
		f, err := os.Open(source)
		if err != nil {
			return nil, err
		}
		defer f.Close()
		r = f
	}

	dec := json.NewDecoder(r)
	dec.UseNumber()
	doc := &settleDoc{}
	if err := dec.Decode(doc); err != nil {
		return nil, err
	}
	return doc, nil
}

func main() {
	source := defaultOrigin
	args := os.Args[1:]
	for i := 0; i < len(args)-1; i++ {
		if args[i] == "--url" || args[i] == "--file" {
			source = args[i+1]
		}
	}

	doc, err := load(source)
	if err != nil {
		fmt.Fprintf(os.Stderr, "could not load vectors from %s: %v\n", source, err)
		os.Exit(2)
	}

	var hashFailures []string
	for _, fixture := range doc.Canonicalisation.Fixtures {
		got, err := hashRecord(fixture.Record)
		if err != nil {
			hashFailures = append(hashFailures, fmt.Sprintf("  canonical hash error: %v", err))
			continue
		}
		if got != fixture.SHA256 {
			hashFailures = append(hashFailures,
				fmt.Sprintf("  canonical hash mismatch\n    expected %s\n    got      %s", fixture.SHA256, got))
		}
	}
	if len(hashFailures) > 0 {
		fmt.Println("NOT CONFORMANT - canonicalisation differs:\n")
		fmt.Println(strings.Join(hashFailures, "\n"))
		os.Exit(1)
	}

	now := doc.EvaluatedAtUnix
	var failures []string
	for _, v := range doc.Vectors {
		expected := fmt.Sprintf("%s/%s/%d", v.Expect.Verdict, v.Expect.Burden, v.Expect.SettledMinor)
		gotObj := adjudicate(v.Chain, now)
		got := fmt.Sprintf("%s/%s/%d", gotObj.Verdict, gotObj.Burden, gotObj.SettledMinor)
		if got != expected {
			failures = append(failures,
				fmt.Sprintf("  %s: expected %s, got %s\n    pins: %s", v.ID, expected, got, v.Pins))
		}
	}

	total := len(doc.Vectors)
	if len(failures) > 0 {
		fmt.Printf("NOT CONFORMANT - %d of %d vectors failed:\n\n", len(failures), total)
		fmt.Println(strings.Join(failures, "\n"))
		os.Exit(1)
	}

	fmt.Printf("CONFORMANT - %d/%d vectors and %d hash fixtures passed.\n",
		total, total, len(doc.Canonicalisation.Fixtures))
}
